File: src/functionalStyle/map.java
public class Node<T> implements Map<T> {
    private string key;
    private T value;
    private Node<T> next;

    public Node(string key, T value, Node<T> next) {
        this.key = key;
        this.value = value;
        this.next = next;
    }

    public T find(string key) {
        if (compareTo(key, this.key) == 0)
            return this.value;
        else 
            return this.next.find(key);
        }

    public Map<T> add(string key, T value) {
        return new Node(key, value, this);
    }
}

File: src/map/ocaml/implicitmodule/bin/map.ml
(* file: map.ml
   author: Bob Muller

   CSCI 1103 Computer Science 1 Honors

   This is an implementation of maps using association lists.
*)
type key = string

type 'a t = Empty
          | Node of { key   : key
                    ; value : 'a
                    ; next  : 'a t
                    }

(* empty : 'a t *)
let empty = Empty

(* add : key -> 'a -> 'a t -> 'a t *)
let add key value map = Node {key; value; next = map}

(* find : key - 'a t -> 'a *)
let rec find key map =
  match map with
  | Empty -> failwith "no key"
  | Node { key = key1; value; next} ->
    let result = compare key key1
    in
    (if result = 0 then
       value
     else
       find key map)
